File: 032.Longest Valid Parentheses.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        int count = 0;
        stack<int> save;
        
        for (int i = 0; i < s.length(); i++) {
            if (s[i] == '(') {
                save.push(i);
            }
            if (s[i] == ')') {
                if (!save.empty() && s[save.top()] == '(') {
                    save.pop();
                    count += 2;
                } else {
                    save.push(i);
                }
            }
        }
        
        if (save.empty())
            return count;
        
        int last = s.length() - 1;
        count = 0;
        while (!save.empty()) {
            int tmp = last - save.top();
            last = save.top();
            save.pop();
            if (count < tmp)
                count = tmp - tmp%2;
        }
        
        if (last > count)
            count = last;
        
        return count;
    }
};

File: 032.Longest Valid Parentheses.java
/*
    Given a string containing just the characters '(' and ')', find the length of the longest valid (well-formed) parentheses substring.

    For "(()", the longest valid parentheses substring is "()", which has length = 2.

    Another example is ")()())", where the longest valid parentheses substring is "()()", which has length = 4.
     */

    int longestValidParentheses(String s) {
        int n = s.length();
        int longest = 0;
        Stack<Integer> stack = new Stack<>();
        for (int i = 0; i < n; i++) {
            if (s.charAt(i) == '(') stack.push(i);
            else {
                if (!stack.isEmpty()) {
                    if (s.charAt(stack.peek()) == '(') stack.pop();
                    else stack.push(i);
                }
                else stack.push(i);
            }
        }
        if (stack.isEmpty()) longest = n;
        else {
            int a = n, b = 0;
            while (!stack.isEmpty()) {
                b = stack.peek(); stack.pop();
                longest = Math.max(longest, a-b-1);
                a = b;
            }
            longest = Math.max(longest, a);
        }
        return longest;
    }
